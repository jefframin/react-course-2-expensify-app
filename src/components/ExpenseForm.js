import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props)

    this.state ={
        description: props.expense ? props.expense.description : '',
        note: props.expense ? props.expense.note : '',
        amount: props.expense ? (props.expense.amount / 100).toString() : '',
        created: props.expense ? moment(props.expense.created) : moment(),
        calFocused: false,
        error: false
    }
    }

    onDescriptionChange = (event) => {
        const desc = event.target.value
        this.setState(() => {return {description: desc}})
    }

    onNoteChange = (event) => {
        const note = event.target.value
        this.setState(() => {return {note: note}})
    }

    onAmountChange = (event) => {
        const amount = event.target.value
        if (!amount || amount.match(/^[1-9]\d*(\.\d{0,2})?$/)) {
            this.setState(() => {return {amount: amount}})
        }
    }

    onDateChange = (created) => {
        if (created) {
          this.setState(() => {return {created: created}})
        }
    }

    onCalFocusChange = ({focused}) => {
        this.setState(() => {return {calFocused: focused}})
    }

    onSubmit = (event) => {
        event.preventDefault();
        if (!this.state.description || !this.state.amount) {
          this.setState(() => {return {error: true}})
        }
        else {
          this.setState(() => {return {error: false}})
          this.props.onSubmit({
              description: this.state.description,
              amount: parseFloat(this.state.amount, 10) * 100,
              created: this.state.created.valueOf(),
              note: this.state.note
          })
        }
    }

    render() {
        return (
            <div>
            {this.state.error && <p>please provide both description and amount</p>}
            <form onSubmit={this.onSubmit}>
            <input type="text" 
                   placeholder="Description" 
                   autoFocus
                   value={this.state.description}
                   onChange={this.onDescriptionChange}></input>
            <input type="text" 
                   value={this.state.amount}
                   placeholder="Amount"
                   onChange={this.onAmountChange}></input>
            <SingleDatePicker date={this.state.created}
                              onDateChange={this.onDateChange}
                              focused={this.state.calFocused}
                              numberOfMonths={1}
                              isOutsideRange={() => {return false}}
                              onFocusChange={this.onCalFocusChange}/>
            <textarea placeholder="Add a note for your expense (optional)"
                      value={this.state.note}
                      onChange={this.onNoteChange}></textarea>
            <button>Save Expense</button>
            </form>
            </div>
        );
    }
}