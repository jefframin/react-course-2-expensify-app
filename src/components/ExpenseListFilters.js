import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import { setTextFilterAction, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component {

    state = {
        calFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calFocused) => {
        this.setState(() => {
            return {calFocused: calFocused}
        })
    }

    onTextChange = (event) => {
       this.props.setTextFilter(event.target.value)
    }

    onSortChange = (event) => {
        event.target.value == 'amount' ? this.props.sortByAmount() : this.props.sortByDate()
    }

    render() {

    return (
        <div>
          <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
          <select value={this.props.filters.order} onChange={this.onSortChange} >
            <option value='date'>Date</option>
            <option value='amount'>Amount</option>
          </select>
          <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => {return false}}
                >
          </DateRangePicker>
        </div>
    )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilterAction(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)