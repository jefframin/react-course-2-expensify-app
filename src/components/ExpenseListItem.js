import React from 'react'
import { Link } from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, created}) => {
    return (
        <div>
          <h3>
          <Link to={`/edit/${id}`}>{description}</Link>
          </h3>
          <p>{amount} - {created}</p>
        </div>
    )
}

export default ExpenseListItem