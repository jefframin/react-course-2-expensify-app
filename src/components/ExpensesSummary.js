import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import getVisibleExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

export const ExpensesSummary = (props) => {
    return (
        <div>
        {props.expenseCount > 0 ?
        <h1>
        Viewing {props.expenseCount} {props.expenseCount > 1 ? 'expenses' : 'expense'} totalling {numeral(props.expensesTotal/100).format('$0,0.00')}
        </h1>
        : ''
        }
        </div>
    )
}

const ConnectedExpensesSummary = connect((state) => {
    const selectedExpenses = getVisibleExpenses(state)
    return {
        expenseCount: selectedExpenses.length,
        expensesTotal: selectExpensesTotal(selectedExpenses)
    }

})(ExpensesSummary)

export default ConnectedExpensesSummary
