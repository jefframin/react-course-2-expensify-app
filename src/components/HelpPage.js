import React from 'react';

const HelpPage = () =>{
    return (
        <div>
        this is help component
        </div>
    )
}

export default HelpPage