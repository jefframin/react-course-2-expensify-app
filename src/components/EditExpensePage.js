import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpenseAction } from '../actions/expenses'
import { removeExpenseAction } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit=(expense) => {
            this.props.editExpenseAction(this.props.expense.id, expense)
            this.props.history.push('/')
        }
    onClick=(event) => {
            this.props.removeExpenseAction({id: this.props.expense.id})
            this.props.history.push('/')
        }

    render() {
    return (
        <div>
          <ExpenseForm expense={this.props.expense}
                       onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
        </div>
    )
    }
}

/*
const EditExpensePage = (props) => {
    console.log(props)
    return (
        <div>
          <ExpenseForm expense={props.expense}
                        onSubmit={(expense) => {
            props.dispatch(editExpenseAction(props.expense.id, expense))
            props.history.push('/')
        }}/>
        <button onClick={(event) => {
            props.dispatch(removeExpenseAction({id: props.expense.id}))
            props.history.push('/')
        }}>Remove</button>
        </div>
    )
}
*/

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        }) 
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        editExpenseAction: (id, expense) => dispatch(editExpenseAction(id, expense)),
        removeExpenseAction: ({ id }) => dispatch(removeExpenseAction({ id }))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)