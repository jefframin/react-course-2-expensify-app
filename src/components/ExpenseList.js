import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => {
    return (
        <div>
        {
            props.expenses.length === 0 ? (
                <p>no expenses</p>
            ) : (
                props.expenses.map((expense) => {
                   return <ExpenseListItem key={expense.id} {...expense}/>
                }) 
            )
        }
        </div>

    )
}

const ConnectedExpenseList = connect((state) => {
    return {
        expenses: selectExpenses(state)
    }

})(ExpenseList)

export default ConnectedExpenseList