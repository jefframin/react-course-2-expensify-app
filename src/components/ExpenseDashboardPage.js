import React from 'react';
import ExpenseList from './ExpenseList'
import ExpensesSummary from './ExpensesSummary'
import ExpenseListFilters from './ExpenseListFilters'

const ExpenseDashboardPage = () => {
    return (
        <div>
        <ExpensesSummary />
        <ExpenseListFilters></ExpenseListFilters>
        <ExpenseList></ExpenseList>
        </div>
    )
}

export default ExpenseDashboardPage