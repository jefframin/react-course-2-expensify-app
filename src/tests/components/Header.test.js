import React from 'react'
import { Header } from '../../components/Header'
import { shallow } from 'enzyme'

test('should render header', () => {
    const wrapper = shallow(<Header startLogout={() => {}}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should call startLogout on button click', () => {
    const logoutSpy = jest.fn()
    const wrapper = shallow(<Header startLogout={logoutSpy}/>)
    wrapper.find('button').simulate('click')
    expect(logoutSpy).toHaveBeenCalled()
})
