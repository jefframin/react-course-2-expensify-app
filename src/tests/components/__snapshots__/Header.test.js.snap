// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <p>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      exact={true}
      to="/dashboard"
    >
      Dashboard
    </NavLink>
  </p>
  <p>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      to="/create"
    >
      Create Expense
    </NavLink>
  </p>
  <p>
    <button
      onClick={[Function]}
    >
      Logout
    </button>
  </p>
</header>
`;
